sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java

```diff
@@ -0,0 +1,234 @@
+
+import static com.carrotsearch.randomizedtesting.RandomizedTest.$;
+import static org.hamcrest.Matchers.is;
+
+public class IngestionServiceIntegrationTest extends SQLTransportIntegrationTest {
+
+    private static final String INGESTION_SOURCE_ID = "testing";
+    private static final String UNDER_TEST_INGEST_RULE_NAME = "ingest_data_rule";
+    private static final String DATA_TOPIC = "test";
+    private static final String TARGET_TABLE = "ingest_data_raw";
+
+    private class TestIngestionImplementation implements IngestionImplementation {
+
+        private final IngestionService ingestionService;
+        private final InputFactory inputFactory;
+        private Set<Tuple<Predicate<Row>, IngestRule>> conditionPredicateAndIngestRules;
+        private final ExpressionAnalyzer expressionAnalyzer;
+        private final ExpressionAnalysisContext expressionAnalysisContext;
+        private FieldProvider<Symbol> inputColumnProvider = new FieldProvider<Symbol>() {
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, Operation operation) {
+                return resolveField(qualifiedName, null, operation);
+            }
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> path, Operation operation) {
+                return new InputColumn(0);
+            }
+        };
+
+        TestIngestionImplementation(Functions functions,
+                                    IngestionService ingestionService) {
+            this.conditionPredicateAndIngestRules = Collections.synchronizedSet(new HashSet<>());
+            this.inputFactory = new InputFactory(functions);
+            this.ingestionService = ingestionService;
+            this.expressionAnalysisContext = new ExpressionAnalysisContext();
+            Row args = new RowN($(DATA_TOPIC));
+            ParameterContext ctx = new ParameterContext(args, Collections.emptyList());
+            this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, ctx, inputColumnProvider, null);
+        }
+
+        void registerIngestionImplementation() {
+            ingestionService.registerImplementation(INGESTION_SOURCE_ID, this);
+        }
+
+        void ingestData(String topic, int data) {
+            for (Tuple<Predicate<Row>, IngestRule> predicateAndRuleTuple : conditionPredicateAndIngestRules) {
+                if (predicateAndRuleTuple.v1().test(new Row1(topic))) {
+                    IngestRule ingestRule = predicateAndRuleTuple.v2();
+                    execute("insert into " + ingestRule.getTargetTable() + " values (?);", new Object[]{data});
+                    execute("refresh table " + ingestRule.getTargetTable());
+                }
+            }
+        }
+
+        @Override
+        public void applyRules(Set<IngestRule> rules) {
+            this.conditionPredicateAndIngestRules = Collections.synchronizedSet(new HashSet<>(rules.size()));
```

Comment:
using an AtomicReference for the instance var and just apply it once the new set is build would be better